### ISSUE TITLE
jinja: should not report undefined variable for defined variable

### ISSUE THREAD

**Issue by erikgb** [ansible/ansible-lint#2447]
##### Summary

Ansible-lint reports undefined variable on a defined variable when variable is used in Jinja expression.

##### Issue Type

- Bug Report

##### Ansible and Ansible Lint details

```console (paste below)
ansible --version
ansible [core 2.13.4]
  config file = None
  configured module search path = ['/home/erikbo/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.10/dist-packages/ansible
  ansible collection location = /home/erikbo/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/local/bin/ansible
  python version = 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0]
  jinja version = 3.1.2
  libyaml = True

ansible-lint --version
ansible-lint 6.6.0 using ansible 2.13.4
```

- ansible installation method: pip
- ansible-lint installation method: pip

##### OS / ENVIRONMENT

Not applicable

##### STEPS TO REPRODUCE

Given this simple playbook (`test.yml`):

```yaml
---
- name: Ansible Lint to_nice_yaml bug
  hosts: localhost
  connection: local
  tasks:
    - name: Set test fact
      ansible.builtin.set_fact:
        my_yaml_var:
          clusters:
            - cluster:
                server: foo:443
              name: foo-cluster:443
      register: updated

    - name: Update test fact
      ansible.utils.update_fact:
        updates:
          - path: "my_yaml_var.clusters[0].cluster.server"
            value: "https://bar:6443"
      register: updated

    - name: Debug updated test fact
      ansible.builtin.debug:
        msg: "{{ updated.my_yaml_var | to_yaml }}"
```

##### Desired Behavior

Should not report any lint issues, as the variable is obviously defined.

##### Actual Behavior

Running `ansible-lint test.yml`:

```paste below
WARNING  Overriding detected file kind 'yaml' with 'playbook' for given positional argument: test.yml
WARNING  Listing 1 violation(s) that are fatal
jinja: to_yaml - 'updated' is undefined. 'updated' is undefined (jinja[invalid])
test.yml:22 Task/Handler: Debug updated test fact

You can skip specific rules or tags by adding them to your configuration file:
# .config/ansible-lint.yml
warn_list:  # or 'skip_list' to silence them completely
  - jinja[invalid]  # Rule that looks inside jinja2 templates.

Finished with 1 failure(s), 0 warning(s) on 1 files.
```

Running the playbook works fine: `ansible-playbook test.yml`

```paste
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [Ansible Lint to_nice_yaml bug] ***************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] *****************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Set test fact] *******************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Update test fact] ****************************************************************************************************************************************************************************************************************************************
changed: [localhost]

TASK [Debug updated test fact] *********************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "clusters:\n- cluster: {server: 'https://bar:6443'}\n  name: foo-cluster:443\n"
}

PLAY RECAP *****************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=4    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

**Comment by briantist** [comment]
I'm seeing what seems like the same behavior when using `register:`ed variables in jinja tests, for example:

```
---
- name: Ansible Lint to_nice_yaml bug
  hosts: localhost
  connection: local
  tasks:
    - name: Set test fact
      ansible.builtin.set_fact:
        my_yaml_var:
          clusters:
            - cluster:
                server: foo:443
              name: foo-cluster:443
      register: updated

    - name: Passes
      when: updated is changed
      ansible.builtin.debug:
        msg: "hi"

    - name: Fails
      vars:
        has_changed: "{{ updated is changed }}"  # fails
      ansible.builtin.debug:
        msg: "{{ updated is changed }}"  # also fails
```

This shows up as:
```
jinja: The 'changed' test expects a dictionary (jinja[invalid])
```

If I change the references to like `{{ updated | default({}) is changed }}` then they will work, but that's ugly and confusing.

Clearly ansible-lint knows that it's ok when used in the `when` keyword, but if it's referenced in the task call, even implicitly via vars, it will fail.

**Comment by ssbarnea** [comment]
I think that https://github.com/ansible/ansible-lint/releases/tag/v6.6.1 is also fixing this as it caused in some cases `jinja[invalid]`.

**Comment by erikgb** [comment]
Seems to work with version 6.6.1! Thanks @ssbarnea!

**Comment by czqrny** [comment]
this bug appeard again in versions 6.13.0 and 6.13.1

**Comment by bendem** [comment]
This still happens in 6.14 for `is changed`:

```console
> cat test.yml          
- name: "`is changed` test expects a dictionary"
  hosts: localhost
  gather_facts: false
  tasks:
    - name: Debug
      ansible.builtin.debug:
        msg: hi
      changed_when: true
      register: output

    - name: Debug 2
      ansible.builtin.debug:
        msg: "{{ output is changed }}"

> ansible-lint --version                 
ansible-lint 6.14.0 using ansible 2.14.3

> ansible-lint -c /dev/null test.yml -vvv
DEBUG    Logging initialized to level 10
DEBUG    Options: Namespace(cache_dir='/home/demarteaub/.cache/ansible-compat/0cb87f', colored=True, configured=True, cwd=PosixPath('/home/demarteaub/projects/infra/ansible'), display_relative_path=True, exclude_paths=['.cache', '.git', '.hg', '.svn', '.tox'], format=None, lintables=['test.yml'], list_rules=False, list_tags=False, write_list=None, parseable=False, quiet=0, rulesdirs=['/home/demarteaub/.local/share/virtualenvs/ansible-JeMCc-2q/lib64/python3.9/site-packages/ansiblelint/rules'], skip_list=[], tags=[], verbosity=3, warn_list=['experimental', 'jinja[spacing]'], kinds=[{'jinja2': '**/*.j2'}, {'jinja2': '**/*.j2.*'}, {'yaml': '.github/**/*.{yaml,yml}'}, {'text': '**/templates/**/*.*'}, {'execution-environment': '**/execution-environment.yml'}, {'ansible-lint-config': '**/.ansible-lint'}, {'ansible-lint-config': '**/.config/ansible-lint.yml'}, {'ansible-navigator-config': '**/ansible-navigator.{yaml,yml}'}, {'inventory': '**/inventory/**.{yaml,yml}'}, {'requirements': '**/meta/requirements.{yaml,yml}'}, {'galaxy': '**/galaxy.yml'}, {'reno': '**/releasenotes/*/*.{yaml,yml}'}, {'tasks': '**/tasks/**/*.{yaml,yml}'}, {'rulebook': '**/rulebooks/*.{yml,yaml'}, {'playbook': '**/playbooks/*.{yml,yaml}'}, {'playbook': '**/*playbook*.{yml,yaml}'}, {'role': '**/roles/*/'}, {'handlers': '**/handlers/*.{yaml,yml}'}, {'vars': '**/{host_vars,group_vars,vars,defaults}/**/*.{yaml,yml}'}, {'test-meta': '**/tests/integration/targets/*/meta/main.{yaml,yml}'}, {'meta': '**/meta/main.{yaml,yml}'}, {'meta-runtime': '**/meta/runtime.{yaml,yml}'}, {'arg_specs': '**/meta/argument_specs.{yaml,yml}'}, {'yaml': '.config/molecule/config.{yaml,yml}'}, {'requirements': '**/molecule/*/{collections,requirements}.{yaml,yml}'}, {'yaml': '**/molecule/*/{base,molecule}.{yaml,yml}'}, {'requirements': '**/requirements.{yaml,yml}'}, {'playbook': '**/molecule/*/*.{yaml,yml}'}, {'yaml': '**/{.ansible-lint,.yamllint}'}, {'changelog': '**/changelogs/changelog.yaml'}, {'yaml': '**/*.{yaml,yml}'}, {'yaml': '**/.*.{yaml,yml}'}, {'sanity-ignore-file': '**/tests/sanity/ignore-*.txt'}], mock_filters=[], mock_modules=[], mock_roles=[], loop_var_prefix=None, only_builtins_allow_collections=[], only_builtins_allow_modules=[], var_naming_pattern=None, offline=None, project_dir='.', extra_vars=None, enable_list=[], skip_action_validation=True, strict=False, rules={}, profile=None, task_name_prefix='{stem} | ', sarif_file=None, list_profiles=False, progressive=False, rulesdir=[], use_default_rules=False, generate_ignore=False, config_file='/dev/null', ignore_file=None, version=False, cache_dir_lock=<filelock._unix.UnixFileLock object at 0x7f685da64280>)
DEBUG    /home/demarteaub/projects/infra/ansible
DEBUG    Loading custom .yamllint.yml config file, this extends our internal yamllint config.
DEBUG    Effective yamllint rules used: {'braces': {'level': 'error', 'forbid': False, 'min-spaces-inside': 1, 'max-spaces-inside': 1, 'min-spaces-inside-empty': 0, 'max-spaces-inside-empty': 0}, 'brackets': {'level': 'error', 'forbid': False, 'min-spaces-inside': 1, 'max-spaces-inside': 1, 'min-spaces-inside-empty': 0, 'max-spaces-inside-empty': 0}, 'colons': {'level': 'error', 'max-spaces-before': 0, 'max-spaces-after': 1}, 'commas': {'level': 'error', 'max-spaces-before': 0, 'min-spaces-after': 1, 'max-spaces-after': 1}, 'comments': {'level': 'warning', 'require-starting-space': True, 'ignore-shebangs': True, 'min-spaces-from-content': 2}, 'comments-indentation': False, 'document-end': False, 'document-start': False, 'empty-lines': {'level': 'error', 'max': 2, 'max-start': 0, 'max-end': 0}, 'empty-values': False, 'float-values': False, 'hyphens': {'level': 'error', 'max-spaces-after': 1}, 'indentation': {'level': 'error', 'spaces': 2, 'indent-sequences': 'consistent', 'check-multi-line-strings': False}, 'key-duplicates': {'level': 'error'}, 'key-ordering': False, 'line-length': False, 'new-line-at-end-of-file': False, 'new-lines': {'level': 'error', 'type': 'unix'}, 'octal-values': False, 'quoted-strings': {'quote-type': 'double', 'required': False, 'allow-quoted-quotes': True, 'level': 'error', 'extra-required': [], 'extra-allowed': []}, 'trailing-spaces': False, 'truthy': False}
INFO     Set ANSIBLE_LIBRARY=/home/demarteaub/.cache/ansible-compat/0cb87f/modules:/home/demarteaub/.ansible/plugins/modules:/usr/share/ansible/plugins/modules
INFO     Set ANSIBLE_COLLECTIONS_PATH=/home/demarteaub/.cache/ansible-compat/0cb87f/collections:/home/demarteaub/projects/infra/ansible/vendors:/home/demarteaub/projects/infra/ansible
INFO     Set ANSIBLE_ROLES_PATH=/home/demarteaub/.cache/ansible-compat/0cb87f/roles:roles:/home/demarteaub/projects/infra/ansible/roles
INFO     Executing syntax check on playbook test.yml (0.45s)
DEBUG    Examining test.yml of type playbook
WARNING  Listing 1 violation(s) that are fatal
jinja[invalid]: The 'changed' test expects a dictionary
test.yml:13 Task/Handler: Debug 2

DEBUG    Attempting to release lock 140086224503424 on /home/demarteaub/.cache/ansible-compat/0cb87f/.lock
DEBUG    Lock 140086224503424 released on /home/demarteaub/.cache/ansible-compat/0cb87f/.lock
Read documentation for instructions on how to ignore specific rule violations.

DEBUG    Determined rule-profile order: {'internal-error': (0, 'min'), 'load-failure': (1, 'min'), 'parser-error': (2, 'min'), 'syntax-check': (3, 'min'), 'command-instead-of-module': (4, 'basic'), 'command-instead-of-shell': (5, 'basic'), 'deprecated-bare-vars': (6, 'basic'), 'deprecated-local-action': (7, 'basic'), 'deprecated-module': (8, 'basic'), 'inline-env-var': (9, 'basic'), 'key-order': (10, 'basic'), 'literal-compare': (11, 'basic'), 'jinja': (12, 'basic'), 'no-free-form': (13, 'basic'), 'no-jinja-when': (14, 'basic'), 'no-tabs': (15, 'basic'), 'partial-become': (16, 'basic'), 'playbook-extension': (17, 'basic'), 'role-name': (18, 'basic'), 'schema': (19, 'basic'), 'name': (20, 'basic'), 'var-naming': (21, 'basic'), 'yaml': (22, 'basic'), 'name[template]': (23, 'moderate'), 'name[imperative]': (24, 'moderate'), 'name[casing]': (25, 'moderate'), 'spell-var-name': (26, 'moderate'), 'avoid-implicit': (27, 'safety'), 'latest': (28, 'safety'), 'package-latest': (29, 'safety'), 'risky-file-permissions': (30, 'safety'), 'risky-octal': (31, 'safety'), 'risky-shell-pipe': (32, 'safety'), 'galaxy': (33, 'shared'), 'ignore-errors': (34, 'shared'), 'layout': (35, 'shared'), 'meta-incorrect': (36, 'shared'), 'meta-no-info': (37, 'shared'), 'meta-no-tags': (38, 'shared'), 'meta-video-links': (39, 'shared'), 'meta-version': (40, 'shared'), 'meta-runtime': (41, 'shared'), 'no-changed-when': (42, 'shared'), 'no-changelog': (43, 'shared'), 'no-handler': (44, 'shared'), 'no-relative-paths': (45, 'shared'), 'max-block-depth': (46, 'shared'), 'max-tasks': (47, 'shared'), 'unsafe-loop': (48, 'shared'), 'avoid-dot-notation': (49, 'production'), 'sanity': (50, 'production'), 'fqcn': (51, 'production'), 'import-task-no-when': (52, 'production'), 'meta-no-dependencies': (53, 'production'), 'single-entry-point': (54, 'production'), 'use-loop': (55, 'production')}
              Rule Violation Summary               
 count tag            profile rule associated tags 
     1 jinja[invalid] basic   formatting           

Failed after min profile: 1 failure(s), 0 warning(s) on 1 files.
```

**Comment by mprasil** [comment]
@bendem I've reported this issue with some steps to reproduce: #3155
